validation.go

```diff
@@ -0,0 +1,383 @@
+package pubsub
+
+import (
+	"context"
+	"fmt"
+	"runtime"
+	"time"
+
+	peer "github.com/libp2p/go-libp2p-peer"
+)
+
+const (
+	defaultValidateConcurrency = 1024
+	defaultValidateThrottle    = 8192
+)
+
+// Validator is a function that validates a message.
+type Validator func(context.Context, peer.ID, *Message) bool
+
+// ValidatorOpt is an option for RegisterTopicValidator.
+type ValidatorOpt func(addVal *addValReq) error
+
+// validation represents the validator pipeline
+type validation struct {
+	p *PubSub
+
+	// topicVals tracks per topic validators
+	topicVals map[string]*topicVal
+
+	// validateQ is the front-end to the validation pipeline
+	validateQ chan *validateReq
+
+	// validateThrottle limits the number of active validation goroutines
+	validateThrottle chan struct{}
+
+	// this is the number of synchronous validation workers
+	validateWorkers int
+}
+
+// validation requests
+type validateReq struct {
+	vals []*topicVal
+	src  peer.ID
+	msg  *Message
+}
+
+// representation of topic validators
+type topicVal struct {
+	topic            string
+	validate         Validator
+	validateTimeout  time.Duration
+	validateThrottle chan struct{}
+	validateInline   bool
+}
+
+// async request to add a topic validators
+type addValReq struct {
+	topic    string
+	validate Validator
+	timeout  time.Duration
+	throttle int
+	inline   bool
+	resp     chan error
+}
+
+// async request to remove a topic validator
+type rmValReq struct {
+	topic string
+	resp  chan error
+}
+
+// newValidation creates a new validation pipeline
+func newValidation() *validation {
+	return &validation{
+		topicVals:        make(map[string]*topicVal),
+		validateQ:        make(chan *validateReq, 32),
+		validateThrottle: make(chan struct{}, defaultValidateThrottle),
+		validateWorkers:  runtime.NumCPU(),
```

Comment:
Could we document somewhere that the defaults assume CPU-intensive validation, and that the worker pool length should be adjusted for IO-bound tasks? Godoc on the `WithValidateWorkers` method should suffice.

Reply:
Well, the validateWorkers perform a cpu-bound task by default, which is signature validation.
But better docs on `WithValidateWorkers` can do.

Reply:
improved the godoc for `WithValidateWorkers`.